File: lib/viewmodels/login_view_model.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:privault/core/app_error.dart';
import 'package:privault/core/base_view_model.dart';
import 'package:privault/core/command_result.dart';
import 'package:privault/database/database.dart';
import 'package:privault/services/biometric_service.dart';
import 'package:privault/services/config_service.dart';
import 'package:privault/services/crypto_service.dart';
import 'package:privault/services/database_service.dart';
import 'package:privault/services/password_service.dart';
import 'package:privault/services/session_service.dart';
import 'package:path/path.dart' as p;
import 'package:uuid/uuid.dart';

/// Das `LoginViewModel` verwaltet den Authentifizierungsprozess und den Zugriff auf die Tresore.
class LoginViewModel extends BaseViewModel {
  // ------------------------------------------------------------------------
  // --- Verwendete Dienste ---
  // ------------------------------------------------------------------------

  final BiometricService _biometricService;
  final ConfigService _configService;
  final CryptoService _cryptoService;
  final DatabaseService _databaseService;
  final PasswordService _passwordService;
  final SessionService _sessionService;

  // ------------------------------------------------------------------------
  // --- Konstanten für notifySuccess/CommandResult ---
  // ------------------------------------------------------------------------

  static const askToEnableBiometrics = -1;

  // ------------------------------------------------------------------------
  // --- Interne Variablen ---
  // ------------------------------------------------------------------------

  String _vaultName = '';
  String _password = '';
  bool _isExists = false;
  bool _hasBiometricKey = false;
  List<String> _existingVaults = [];

  // ------------------------------------------------------------------------
  // --- Initialisierung ---
  // ------------------------------------------------------------------------

  /// Konstruktor
  LoginViewModel(this._biometricService, this._configService, this._cryptoService, this._databaseService, this._passwordService, this._sessionService);

  /// Initialisiert die Variablen, bevor der erste Frame gerendert wird.
  void init() {
    _vaultName = _configService.lastVaultName;
    _password = '';
    _isExists = false;
    _hasBiometricKey = false;
    _existingVaults = [];
  }

  /// Lädt die Daten.
  Future<void> load() async {
    setBusy(true);
    try {
      await refreshVaultList();
    } catch (e, st) {
      logError('Fehler beim Initialisieren: $e', st);
      notifyError(AppError.unknown);
    } finally {
      setBusy(false);
    }
  }

  /// Führt eine Bereinigung durch (z.B. bei korruptem Tresor).
  Future<void> cleanUp() async {
    await _databaseService.deleteCurrentDatabase();
    _sessionService.clearSession();
    _vaultName = '';
    _password = '';
    _configService.lastVaultName = '';
    await refreshVaultList();
  }

  // ------------------------------------------------------------------------
  // --- Login-Button ---
  // ------------------------------------------------------------------------

  /// Startet den Authentifizierungsprozess.
  /// Prüft, ob der Tresor existiert, und leitet dann entweder das Öffnen oder die Neuanlage ein.
  Future<CommandResult<int>> login({bool forceCreate = false}) async {
    _vaultName = _vaultName.trim();

    // Validierung der Benutzereingabe
    if (_vaultName.isEmpty) {
      return notifyError(AppError.valueRequired, field: 'vaultName');
    }
    if (!_isExists && !forceCreate) {
      return notifyError(AppError.vaultNotFound, field: 'vaultName');
    }

    setBusy(true);
    try {
      clearError();

      // Kurze Pause für Lade-Indikator, bevor Argon2 blockiert
      await Future.delayed(const Duration(milliseconds: 50));

      // Sicherstellen, dass keine alte Verbindung mehr offen ist
      await _databaseService.close();

      if (_isExists) {
        return await _openVault();
      } else {
        if (_password.isEmpty) {
          return notifyError(AppError.valueRequired, field: 'password');
        }
        return await _createVault();
      }
    } catch (e, st) {
      final msg = e.toString().toLowerCase();

      if (msg.contains('database is locked')) {
        return notifyError(AppError.vaultLocked);
      }

      if (msg.contains('file is not a database') || msg.contains('authentication failed') || msg.contains('file is encrypted or is not a database')) {
        return notifyError(AppError.wrongPassword, field: 'password');
      }

      logError('Fehler beim Login: $e', st);
      return notifyError(AppError.unknown);
    } finally {
      setBusy(false);
    }
  }

  /// Erstellt eine neue Tresor-Datenbank, generiert ein RSA-Schlüsselpaar für die Identität
  /// und verschlüsselt den privaten Teil mit dem Master-Key.
  Future<CommandResult<int>> _createVault() async {
    if (_password.isEmpty) {
      return notifyError(AppError.valueRequired, field: 'password');
    }

    // 1. Neues Salt generieren
    final salt = _cryptoService.generateSalt();

    // 2. Master-Key ableiten
    final masterKey = await _cryptoService.deriveKey(_password, salt);

    try {
      // Bestehende löschen
      //await _databaseService.deleteCurrentDatabase();

      // 3. Datenbank initialisieren (erstellt Tabellen)
      await _databaseService.initialize(_vaultName, masterKey);

      // 4. Salt-Datei anlegen
      // Alle Zeichen außer Buchstaben, Zahlen, Unterstrichen und Bindestrichen durch '_' ersetzen.
      final safeName = vaultName.replaceAll(RegExp(r'[<>:"/\\|?*]'), '_').trim();
      final saltFile = File(p.join(_configService.vaultStoragePath, '$safeName.db3.salt'));
      await saltFile.writeAsBytes(salt);

      // 5. RSA-Schlüsselpaar für den User generieren
      final (pubKey, privKeyBytes) = await _cryptoService.generateRsaKeyPair();

      // 6. Private Key verschlüsseln (mit dem Master-Key)
      final encryptedPrivKey = await _cryptoService.encrypt(privKeyBytes, masterKey);

      // 7. UserEntity mit der ID = 1 (Benutzer der App) erstellen.
      // SQLite-net schaut beim Insert in seine interne Sequenz-Tabelle.
      // Da die Datenbank neu ist, ist die nächste freie ID immer die 1.
      final newUser = await _databaseService.saveUser(UserEntity(
        id: 0,
        uuid: const Uuid().v4(),
        name: Platform.environment['USERNAME'] ?? 'User',
        publicKey: pubKey,
        isVerified: true,
        isHidden: false,
        updatedAt: DateTime.now().toUtc(),
      ));

      // 8. Settings anlegen
      final newSettings = await _databaseService.saveSettings(SettingsEntity(
        id: 0,
        salt: base64.encode(salt),
        encryptedPrivateKey: encryptedPrivKey,
        host: kDebugMode ? 'https://privault.test/api' : '', // todo später wieder auskommentieren!!!!
        apiToken: kDebugMode ? '6h54qT5l2r37Kr7XxfP08YD7gPAGff6aWSaa' : '', // todo später wieder auskommentieren!!!!
        useBiometric: false,
        pwLength: 16,
        pwSpecialChars: '',
        pwAvoidIlO0: true,
        categoryPlaceholder: '',
        lastSyncAt: DateTime.fromMillisecondsSinceEpoch(0).toUtc(),
      ));

      // 9. Letzten Tresor merken
      _configService.lastVaultName = _vaultName;
      await refreshVaultList();

      // 10. Session setzen
      _sessionService.setSession(
        user: newUser,
        privateKey: privKeyBytes,
        vaultName: _vaultName,
        settings: newSettings,
      );

      clearPassword();
      return notifySuccess();
    } finally {
      // Master-Key sofort aus dem RAM löschen
      _cryptoService.wipeKey(masterKey);
    }
  }

  /// Öffnet einen bestehenden Tresor, leitet den Master-Key ab und entschlüsselt die Sitzungsdaten.
  Future<CommandResult<int>> _openVault() async {
    // Salt aus der Salt-Datei lesen
    final salt = await _databaseService.getSalt(_vaultName);
    if (salt == null) return notifyError(AppError.vaultNotFound);

    Uint8List? masterKey;
    bool isManualLogin = _password.isNotEmpty;

    // 1. Master-Key aus dem Secure-Store holen oder aus dem eingegebenen Master-Passwort ableiten
    if (!isManualLogin && _hasBiometricKey) {
      // Master-Key aus Secure-Store holen (löst System-Dialog aus)
      masterKey = await _biometricService.getMasterKey(_vaultName);
      if (masterKey == null) {
        return notifyError(AppError.biometricCanceled);
      }
    } else {
      // Master-Key aus eingegebenes Master-Passwort und Salt ableiten (Argon2id)
      if (_password.isEmpty) {
        return notifyError(AppError.valueRequired, field: 'password');
      }
      masterKey = await _cryptoService.deriveKey(_password, salt);
    }

    try {
      // 2. Datenbank öffnen
      await _databaseService.initialize(_vaultName, masterKey);

      // 3. Einstellungen und eigene Identität aus der DB lesen
      final settings = await _databaseService.getSettings();
      final user = await _databaseService.getUser(1);
      if (settings == null || user == null) return notifyError(AppError.vaultCorrupt);

      // 4. Private Key entschlüsseln
      try {
        final privKeyBytes = await _cryptoService.decrypt(settings.encryptedPrivateKey, masterKey);
        _sessionService.setSession(user: user, privateKey: privKeyBytes, vaultName: _vaultName, settings: settings);
      } catch (e) {
        if (!isManualLogin && _hasBiometricKey) {
          await _biometricService.removeMasterKey(_vaultName);
          _hasBiometricKey = false;
          return notifyError(AppError.wrongBiometric, field: 'password');
        }
        return notifyError(AppError.wrongPassword, field: 'password');
      }

      _configService.lastVaultName = _vaultName;
      clearPassword();

      // Falls manuell eingeloggt und Biometrie gewünscht, aber noch nicht hinterlegt: Nachfragen
      if (isManualLogin && settings.useBiometric && !_hasBiometricKey) {
        return notifySuccess(askToEnableBiometrics);
      }

      return notifySuccess();
    } catch (e) {
      await _databaseService.close();
      rethrow; // Wird im login() gefangen
    } finally {
      // Hygiene: Master-Key aus dem RAM löschen
      _cryptoService.wipeKey(masterKey);
    }
  }


  // ------------------------------------------------------------------------
  // --- Eigenschaften und Methoden ---
  // ------------------------------------------------------------------------

  // --- Tresor ---

  /// Der Name des Tresors, der geöffnet oder neu erstellt werden soll.
  String get vaultName => _vaultName;

  set vaultName(String value) {
    if (_vaultName == value) return;
    // Ungültige Zeichen für Dateinamen filtern
    _vaultName = value.replaceAll(RegExp(r'[<>:"/\\|?*]'), '_');
    clearFieldError('vaultName');
    _updateState();
  }

  /// Gibt an, ob der gewählte Tresor bereits lokal existiert.
  bool get isExists => _isExists;

  /// Eine Liste aller auf diesem Gerät gefundenen Tresore.
  List<String> get existingVaults => _existingVaults;

  /// Scannt das Dateisystem nach vorhandenen Tresor-Datenbanken.
  Future<void> refreshVaultList() async {
    final path = _configService.vaultStoragePath;
    if (path.isEmpty) {
      _existingVaults = [];
    } else {
      final dir = Directory(path);
      if (await dir.exists()) {
        final List<String> vaults = [];
        await for (final file in dir.list()) {
          if (file.path.endsWith('.db3.salt')) {
            final baseName = p.basename(file.path).replaceAll('.db3.salt', '');
            final dbFile = File(p.join(path, '$baseName.db3'));
            if (await dbFile.exists()) vaults.add(baseName);
          }
        }
        _existingVaults = vaults;
      }
    }
    await _updateState();
  }

  /// Prüft, ob es den Tresor gibt und ob es für diesen Tresor ein Wert im Secure-Store liegt.
  Future<void> _updateState() async {
    _isExists = _existingVaults.contains(_vaultName);
    _hasBiometricKey = await _biometricService.containsMasterKey(_vaultName);
    notifyListeners();
  }

  // --- Passwort ---

  /// Das aktuell eingegebene Master-Passwort.
  String get password => _password;

  set password(String value) {
    if (_password == value) return;
    _password = value;
    clearFieldError('password');
    notifyListeners();
  }

  /// Berechnete Stärke des aktuell eingegebenen Passworts (0-4).
  int get passwordStrength => _passwordService.estimateStrength(_password);

  /// Löscht das Passwort aus dem RAM.
  void clearPassword({bool notify = true}) {
    _password = '';
    if (notify) notifyListeners();
  }

  // --- Biometrie ---

  /// Gibt an, ob für den aktuell gewählten Tresor der Master-Key im Secure-Store liegt.
  bool get hasBiometricKey => _hasBiometricKey;

  /// Speichert den abgeleiteten Master-Key im biometrisch geschützten Secure-Store des Betriebssystems.
  Future<void> saveMasterKey(String password) async {
    final salt = await _databaseService.getSalt(_vaultName);
    if (salt == null) return;

    final masterKey = await _cryptoService.deriveKey(password, salt);
    try {
      await _biometricService.saveMasterKey(_vaultName, masterKey);
      await _updateState();
    } finally {
      _cryptoService.wipeKey(masterKey);
    }
  }
}

File: old/Privault.Code/ViewModels/LoginViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Privault.Core.Models.Entities;
using Privault.Core.Services.Contracts;

namespace Privault.Core.ViewModels;

/// <summary>
/// Das <see cref="LoginViewModel"/> verwaltet den Authentifizierungsprozess und den Zugriff auf die Tresore.
/// <para>
/// <b>Hauptaufgaben:</b>
/// <list type="bullet">
/// <item>Anzeige und Auswahl vorhandener lokaler Tresore.</item>
/// <item>Erstellen / Öffnen des Tresors inklusive Schlüsselableitung (Argon2id).</item>
/// <item>Biometrie-Unterstützung: Kein Passwort → Login per Fingerabdruck oder Gesichtserkennung.</item>
/// </list>
/// </para>
/// <para>
/// <b>Sicherheitsaspekte:</b>
/// <list type="bullet">
/// <item>Kein Master-Passwort im RAM: Das Passwort wird nur kurzzeitig zur Ableitung des Master-Keys verwendet.</item>
/// <item>Wiping: Der Master-Key wird nach der Entschlüsselung sofort aus dem Speicher gelöscht (<see cref="ICryptoService.WipeKey"/>).</item>
/// <item>Hardware-Schutz für Biometrie: Der Master-Key liegt im verschlüsselten Secure-Store des Betriebssystems.</item>
/// </list>
/// </para>
/// </summary>
public partial class LoginViewModel : ObservableObject
{
    // ------------------------------------------------------------------------
    // --- Felder ---
    // ------------------------------------------------------------------------

    private readonly IBiometricService _biometricService;
    private readonly IConfigService _configService;
    private readonly ICryptoService _cryptoService;
    private readonly IDatabaseService _databaseService;
    private readonly ISessionService _sessionService;
    private readonly IUiService _uiService;

    /// <summary>
    /// Gibt an, ob Biometrie von der Hardware unterstützt wird und in der Systemeinstellung aktiviert ist.
    /// </summary>
    private bool _isBiometricAvailable;
    
    /// <summary>
    /// Signalisiert, ob gerade die Seite initialisiert wird.
    /// </summary>
    private bool _isInit;

    // ------------------------------------------------------------------------
    // --- Konstruktor ---
    // ------------------------------------------------------------------------

    /// <summary>
    /// Initialisiert eine neue Instanz des <see cref="LoginViewModel"/>.
    /// </summary>
    /// <param name="biometricService">Dienst für die Biometrie-Unterstützung.</param>
    /// <param name="configService">Dienst für globale App-Konfigurationen.</param>
    /// <param name="cryptoService">Dienst für kryptografische Funktionen.</param>
    /// <param name="databaseService">Dienst für den Datenbankzugriff (SQLite/SQLCipher).</param>
    /// <param name="sessionService">Dienst zur Verwaltung der aktuellen Sitzungsdaten (RAM).</param>
    /// <param name="uiService">UI-Abstraktion (Dialoge, Navigation, Toast).</param>
    /// <exception cref="ArgumentNullException">Wird geworfen, wenn einer der benötigten Services null ist.</exception>
    public LoginViewModel(
        IBiometricService biometricService,
        IConfigService configService,
        ICryptoService cryptoService,
        IDatabaseService databaseService,
        ISessionService sessionService,
        IUiService uiService)
    {
        _biometricService = biometricService ?? throw new ArgumentNullException(nameof(biometricService));
        _configService = configService ?? throw new ArgumentNullException(nameof(configService));
        _cryptoService = cryptoService ?? throw new ArgumentNullException(nameof(cryptoService));
        _databaseService = databaseService ?? throw new ArgumentNullException(nameof(databaseService));
        _sessionService = sessionService ?? throw new ArgumentNullException(nameof(sessionService));
        _uiService = uiService ?? throw new ArgumentNullException(nameof(uiService));

        // Letzten Tresor vorausfüllen, Fallback: Benutzername des Betriebssystems
        var lastVault = _configService.LastVaultName;
        VaultName = !string.IsNullOrEmpty(lastVault) ? lastVault : Environment.UserName;
    }
    
    /// <summary>
    /// Initialisiert das ViewModel, nachdem die View geladen wurde.
    /// </summary>
    /// <remarks>
    /// Dieser Befehl wird über die View (per Code-Behind) aufgerufen.
    /// </remarks>
    [RelayCommand]
    private async Task InitializeAsync()
    {
        if (_isInit || IsBusy) return;
        _isInit = true;
        try
        {
            _isBiometricAvailable = await _biometricService.IsAvailableAsync();
            await UpdateStateAsync();
        }
        catch (Exception ex)
        {
            await _uiService.ErrorAsync(ex.Message);
        }
        finally
        {
            _isInit = false;
        }
    }
    
    // ------------------------------------------------------------------------
    // --- Eigenschaften ---
    // ------------------------------------------------------------------------

    /// <summary>
    /// Der Name des Tresors, der geöffnet oder neu erstellt werden soll.
    /// Sonderzeichen werden automatisch durch Unterstriche ersetzt.
    /// </summary>
    [ObservableProperty] 
    private string _vaultName = string.Empty;
    
    /// <summary>
    /// Gibt an, ober der Tresor existiert
    /// </summary>
    [ObservableProperty] 
    private bool _isExists;
    
    /// <summary>
    /// Gibt an, ob für den aktuell gewählten Tresor der Master-Key im Secure-Store liegt.
    /// </summary>
    [ObservableProperty]
    [NotifyCanExecuteChangedFor(nameof(LoginCommand))]
    private bool _hasBiometricKey;
    
    /// <summary>
    /// Das eingegebene Master-Passwort für den Login.
    /// </summary>
    [ObservableProperty] 
    [NotifyCanExecuteChangedFor(nameof(LoginCommand))] 
    private string _password = string.Empty;
    
    /// <summary>
    /// Gibt an, ob aktuell ein langlaufender Prozess (z.B. Argon2-Schlüsselableitung) aktiv ist.
    /// Deaktiviert automatisch den Login-Button.
    /// </summary>
    [ObservableProperty] 
    [NotifyCanExecuteChangedFor(nameof(LoginCommand))]
    private bool _isBusy;

    /// <summary>
    /// Signalisiert verzögert, ob ein Ladevorgang aktiv ist.
    /// </summary>
    [ObservableProperty] 
    private bool _showBusyIndicator;
    
    // ------------------------------------------------------------------------
    // --- Befehle ---
    // ------------------------------------------------------------------------
    
    /// <summary>
    /// Startet den Authentifizierungsprozess.
    /// <para>
    /// Prüft, ob der Tresor existiert, und leitet dann entweder das Öffnen oder die Neuanlage ein.
    /// </para>
    /// </summary>
    [RelayCommand(CanExecute = nameof(CanLogin))]
    private async Task LoginAsync()
    {
        if (_isInit || IsBusy) return;
        IsBusy = true;
        try
        {
            if (IsExists)
            {
                await OpenVaultAsync();
            }
            else
            {
                await CreateVaultAsync();
            }
            
            // // Nach erfolgreichem Login "warmup" triggern
            // _ = Task.Run(() => 
            // {
            //     // 1. JSON Reflektion aufwärmen
            //     var dummyJson = System.Text.Json.JsonSerializer.Serialize(new EntryPayload { Title = "Warmup" });
            //     _ = System.Text.Json.JsonSerializer.Deserialize<EntryPayload>(dummyJson);
            //
            //     // 2. Krypto-Provider initialisieren
            //     _ = _cryptoService.ComputeHash("warmup");
            // });
        }
        catch (Exception ex)
        {
            await _uiService.ErrorAsync(ex.Message);
            await _databaseService.CloseConnectionAsync();
        }
        finally
        {
            IsBusy = false;
        }
    }
    
    /// <summary>
    /// Öffnet ein Menü zur Auswahl eines bereits auf dem Gerät vorhandenen Tresors.
    /// </summary>
    [RelayCommand]
    private async Task PickVaultAsync()
    {
        var vaultMap = _configService.Vaults;
        if (vaultMap.Count == 0)
        {
            await _uiService.InfoAsync("Es sind noch keine lokalen Tresore vorhanden.");
            return;
        }

        var vaults = vaultMap.Keys.ToArray();
        var result = await _uiService.ActionSheetAsync(
            "Tresor wählen", 
            "Abbrechen", 
            null,
            vaults);

        if (!string.IsNullOrWhiteSpace(result) && result != "Abbrechen")
        {
            VaultName = result;
            // Passwort-Feld fokussieren
            OnVaultPicked?.Invoke();
        }
    }
    
    // ------------------------------------------------------------------------
    // --- Ereignishandler ---
    // ------------------------------------------------------------------------

    // /// <summary>
    // /// Wird aufgerufen, wenn IsBusy geändert wird.
    // /// Schaltet den Busy-Indikator verzögert ein bzw. sofort aus.
    // /// </summary>
    // /// <param name="value"></param>
    // partial void OnIsBusyChanged(bool value)
    // {
    //     if (value)
    //         _ = DelayBusyIndicatorAsync(); // fire-and-forget
    //     else
    //         ShowBusyIndicator = false;
    // }
    
    /// <summary>
    /// Wird aufgerufen, wenn sich der VaultName ändert, um ungültige Zeichen sofort zu entfernen.
    /// </summary>
    partial void OnVaultNameChanged(string value)
    {
        if (string.IsNullOrEmpty(value)) return;

        // Ungültige Zeichen entfernen.
        var invalidChars = Path.GetInvalidFileNameChars();
        var sanitized = string.Join("_", value.Split(invalidChars, StringSplitOptions.None));
        if (sanitized != value)
            VaultName = sanitized;
        
        // Status-Flags aktualisieren 
        _ = UpdateStateAsync(); // fire-and-forget      
    }
    
    /// <summary>
    /// Tritt auf, wenn ein Tresor aus der Liste gewählt wurde. 
    /// Die View kann darauf reagieren, um z.B. den Fokus in das Passwortfeld zu setzen.
    /// </summary>
    public event Action? OnVaultPicked;
    
    // ------------------------------------------------------------------------
    // --- Private Methoden ---
    // ------------------------------------------------------------------------

    // --- Kern-Logik ---
    
    /// <summary>
    /// Erstellt eine neue Tresor-Datenbank, generiert ein RSA-Schlüsselpaar für die Identität 
    /// und verschlüsselt den privaten Teil mit dem Master-Key.
    /// </summary>
    private async Task CreateVaultAsync()
    {
        // Entweder war der Tresor nie da, oder er wurde gerade wegen Inkonsistenz gelöscht.
        var createNew = await _uiService.ConfirmAsync(
            "Tresor anlegen",
            $"Der Tresor '{VaultName}' existiert auf diesem Gerät noch nicht.\nMöchtest du ihn anlegen?",
            "Ja, anlegen",
            "Nein, abbrechen");

        if (!createNew) return;
        
        // 1. Neues Salt generieren
        var salt = _cryptoService.GenerateSalt();
        var saltBase64 = Convert.ToBase64String(salt);

        // 2. Master-Key ableiten
        var masterKey = await _cryptoService.DeriveKeyAsync(Password, salt);

        try
        {
            // 3. Datenbank initialisieren (erstellt Tabellen)
            await _databaseService.InitializeAsync(VaultName, masterKey);

            // 4. Salt in die Map eintragen
            var map = _configService.Vaults;
            map[VaultName] = saltBase64;
            _configService.Vaults = map;

            // 5. RSA-Schlüsselpaar für den User generieren
            var (pubKey, privKey) = _cryptoService.GenerateRsaKeyPair();

            // 6. Private Key verschlüsseln (mit dem Master-Key)
            var encryptedPrivKey = _cryptoService.Encrypt(privKey, masterKey);

            // 7. UserEntity mit der ID = 1 (Benutzer der App) erstellen. 
            // SQLite-net schaut beim Insert in seine interne Sequenz-Tabelle.
            // Da die Datenbank neu ist, ist die nächste freie ID immer die 1.
            var newUser = new UserEntity
            {
                //Id = 1, // wird automatisch gesetzt 
                Uuid = Guid.NewGuid().ToString(), // Direkt lokal generieren (V4)   
                Name = Environment.UserName, // muss eindeutig im Tresor sein
                PublicKey = pubKey,
                IsVerified = true, // Ich vertraue mir selbst :-)
            };

            var newSettings = new SettingsEntity
            {
                Salt = saltBase64,
                EncryptedPrivateKey = encryptedPrivKey
            };

            // 8. User und Settings in der DB speichern
            await _databaseService.SaveUserAsync(newUser);
            await _databaseService.SaveSettingsAsync(newSettings);

            // 9. Letzten Tresor merken
            _configService.LastVaultName = VaultName;

            // 10. Session setzen (damit wir direkt eingeloggt sind)
            _sessionService.User = newUser;
            _sessionService.VaultName = VaultName;
            _sessionService.Settings = newSettings;
            _sessionService.PrivateKey = privKey;
        }
        finally
        {
            // Master-Key aus dem RAM löschen
            _cryptoService.WipeKey(masterKey);
        }

        // 11. Aufräumen und zur Liste springen
        Password = string.Empty; // GC-Hinweis
        await GoToMainPageAsync();
    }
    
    /// <summary>
    /// Öffnet einen bestehenden Tresor, leitet den Master-Key ab und entschlüsselt die Sitzungsdaten.
    /// <para>
    /// Wenn die Biometrie genutzt wird, muss der Tresor bereits existieren.
    /// </para>
    /// </summary>
    /// <remarks>
    /// Es wird vorausgesetzt, dass der Tresor in der lokalen Tresorliste steht.
    /// </remarks>
    private async Task OpenVaultAsync()
    {
        byte[]? masterKey;
        
        // Biometrie nutzen, wenn das Passwortfeld leer ist UND Master-Key im Secure-Store liegt
        var useBiometrics = string.IsNullOrEmpty(Password) && HasBiometricKey;
        if (useBiometrics)
        {
            // 1. Master-Key aus Secure-Store holen (löst System-Dialog aus)
            masterKey = await _biometricService.GetMasterKeyAsync(VaultName);
            if (masterKey == null)
                throw new Exception("Biometrie abgebrochen oder nicht konfiguriert.");
        }
        else
        {
            // 1. Master-Key aus eingegebenes Master-Passwort und Salt ableiten (Argon2id)
            var saltBase64 = _configService.Vaults[VaultName]; // zum Tresor zugehörigen Salt aus der lokalen Tresorliste laden
            var salt = Convert.FromBase64String(saltBase64);
            masterKey = await _cryptoService.DeriveKeyAsync(Password, salt);
        }

        try
        {
            UserEntity? user;
            SettingsEntity? settings;
            try
            {
                // 2. Datenbank öffnen
                await _databaseService.InitializeAsync(VaultName, masterKey);

                if (VaultName == "frank" && Password == "falsch") // TODO Hack wieder rausnehmen, ist nur zum Test!!!!!!!!!!
                    throw new Exception("file is not a database");
                
                // 3. Eigene Identität (öffentlich) und Einstellungen (privat) auslesen
                user = await _databaseService.GetUserAsync(1); // Ich bin immer ID 1
                settings = await _databaseService.GetSettingsAsync();
            }
            catch (Exception ex)
            {
                if (ex.Message.Contains("file is not a database"))
                {
                    if (useBiometrics)
                        await _biometricService.RemoveMasterKeyAsync(VaultName);
                    throw new Exception("Falsches Master-Passwort.");
                }
                throw;
            }

            // 4. Private Key entschlüsseln
            try
            {
                if (user == null || settings == null) throw new Exception("Tresor ist korrupt.");
                _sessionService.User = user;
                _sessionService.VaultName = VaultName;
                _sessionService.Settings = settings;
                _sessionService.PrivateKey = _cryptoService.Decrypt(settings.EncryptedPrivateKey, masterKey);
            }
            catch (Exception)
            {
                // Tresor ist beschädigt
                var confirm = await _uiService.ConfirmAsync(
                    "Tresor löschen",
                    "Der Tresor ist korrupt. Soll er gelöscht werden?",
                    "Ja, löschen",
                    "Nein, abbrechen");
                if (confirm)
                    await CleanUpAsync();
                return;
            }
            
            // Falls es ein manueller Login war:
            // Wenn Biometrie in den Einstellungen aktiviert ist, aber und noch nichts im Secure-Store liegt,
            // nachfragen, ob der Master-Key nun im Secure-Store abgelegt werden soll.
            if (!useBiometrics && settings.UseBiometric && !await _biometricService.ContainsMasterKeyAsync(VaultName))
            {
                var confirm = await _uiService.ConfirmAsync(
                    "Biometrie",
                    "Soll dein Schlüssel sicher auf diesem Gerät abgelegt werden, damit du dich beim " +
                                "nächsten Mal bequem per Fingerabdruck oder Gesichtserkennung anmelden kannst?",
                    "Ja, Schlüssel speichern",
                    "Nein, Biometrie deaktivieren");
                if (confirm)
                {
                    await _biometricService.SaveMasterKeyAsync(VaultName, masterKey);
                    await UpdateStateAsync();
                }
                else
                {
                    settings.UseBiometric = false;
                    await _databaseService.SaveSettingsAsync(settings);
                }
            }
        }
        finally
        {
            // Master-Key aus dem Arbeitsspeicher entfernen
            _cryptoService.WipeKey(masterKey);
        }

        // 5. Erfolg: Letzten Tresor merken
        _configService.LastVaultName = VaultName;

        // 6. Aufräumen (Passwort-String) und zur Liste springen
        Password = string.Empty;
        await GoToMainPageAsync();
    }
    
    // --- Sonstige Helfer ---
    
    /// <summary>
    /// Validiert, ob die für einen Login notwendigen Felder ausgefüllt sind.
    /// </summary>
    private bool CanLogin()
    {
        if (IsBusy) 
            return false;
        
        if (string.IsNullOrWhiteSpace(VaultName))
            return false; // kein Tresorname

        // Login ist möglich, wenn ein Passwort eingegeben wurde oder der Master-Key im Secure-Store liegt
        return !string.IsNullOrWhiteSpace(Password) || HasBiometricKey;
    }
    
    /// <summary>
    /// Führt eine Notfall-Bereinigung durch, wenn ein Tresor als korrupt oder inkonsistent erkannt wird.
    /// Löscht die Datenbankdatei und den Eintrag in der Konfigurations-Map.
    /// </summary>
    private async Task CleanUpAsync()
    {
        // Datenbank löschen
        await _databaseService.DeleteCurrentDatabase();

        // Lokalen Map-Eintrag entfernen
        var map = _configService.Vaults;
        map.Remove(VaultName);
        _configService.Vaults = map;
        
        // Master-Key aus dem Secure-Store löschen, falls vorhanden
        if (HasBiometricKey)
            await _biometricService.RemoveMasterKeyAsync(VaultName);
        
        // Status-Flags aktualisieren
        await UpdateStateAsync();
    }
    
    /// <summary>
    /// Navigiert zur Hauptansicht und setzt den Navigationsstack zurück.
    /// </summary>
    private async Task GoToMainPageAsync()
    {
        try
        {
            await _uiService.NavigateAsync("/main");
        }
        catch (Exception ex)
        {
            await _uiService.ErrorAsync($"Main-Navigation fehlgeschlagen: {ex.Message}");
        }
    }

    // /// <summary>
    // /// Schaltet im Login-Prozess den Busy-Indikator verzögert ein.
    // /// </summary>
    // private async Task DelayBusyIndicatorAsync()
    // {
    //     await Task.Delay(250);
    //     if (IsBusy)
    //         ShowBusyIndicator = true;
    // }
    
    /// <summary>
    /// Prüft, ob es den Tresor gibt und ob es für diesen Tresor ein Wert im Secure-Store liegt.
    /// </summary>
    private async Task UpdateStateAsync()
    {
        IsExists = !string.IsNullOrWhiteSpace(VaultName) && _configService.Vaults.ContainsKey(VaultName);
        if (_isBiometricAvailable && IsExists)
        {
            HasBiometricKey = await _biometricService.ContainsMasterKeyAsync(VaultName);
        }
        else
        {
            HasBiometricKey = false; // Biometrie ist nicht verfügbar oder Tresor existiert nicht
        }
        LoginCommand.NotifyCanExecuteChanged();
    }
}